mlir/lib/QEC/Transforms/CommuteCliffordTPPR.cpp

```diff
@@ -0,0 +1,176 @@
+#include "QEC/IR/QECOpInterfaces.h"
+#include "QEC/Transforms/Patterns.h"
+#include "QEC/Utils/PauliStringWrapper.h"
+
+using namespace stim;
+using namespace mlir;
+using namespace catalyst::qec;
+
+namespace {
+
+bool verifyPrevNonClifford(PPRotationOp op, Operation *prevOp)
+{
+    if (prevOp == nullptr)
+        return true;
+
+    if (prevOp == op)
+        return false;
+
+    if (prevOp->isBeforeInBlock(op))
+        return true;
+
+    for (auto operandValue : prevOp->getOperands()) {
+        if (!verifyPrevNonClifford(op, operandValue.getDefiningOp()))
+            return false;
+    }
+    return true;
+}
+
+bool verifyNextNonClifford(PPRotationOp op, PPRotationOp nextOp)
+{
+    if (!nextOp.isNonClifford())
+        return false;
+
+    if (nextOp == nullptr)
+        return false;
+
+    for (auto operandValue : nextOp.getOperands()) {
+        auto defOp = operandValue.getDefiningOp();
+
+        if (defOp == op)
+            continue;
+
+        if (!verifyPrevNonClifford(op, defOp))
+            return false;
+    }
+
+    return true;
+}
+
+LogicalResult visitPPRotationOp(PPRotationOp op,
+                                std::function<LogicalResult(PPRotationOp)> callback)
+{
+    if (op.isNonClifford())
+        return failure();
+
+    for (auto userOp : op->getUsers()) {
+        if (auto pprOp = llvm::dyn_cast<PPRotationOp>(userOp)) {
+            if (verifyNextNonClifford(op, pprOp)) {
+                return callback(pprOp);
+            }
+        }
+    }
+
+    return failure();
+}
+
+void moveCliffordPastNonClifford(const PauliStringWrapper &lhsPauli,
+                                 const PauliStringWrapper &rhsPauli, PauliStringWrapper *result,
+                                 PatternRewriter &rewriter)
+{
+    assert(lhsPauli.op != nullptr && "LHS Operation is not found");
+    assert(rhsPauli.op != nullptr && "RHS Operation is not found");
+    assert(llvm::isa<PPRotationOp>(lhsPauli.op) && "LHS Operation is not PPRotationOp");
+    assert(llvm::isa<PPRotationOp>(rhsPauli.op) && "RHS Operation is not PPRotationOp");
+
+    auto lhs = llvm::cast<PPRotationOp>(lhsPauli.op);
+    auto rhs = llvm::cast<PPRotationOp>(rhsPauli.op);
+
+    assert(lhs.isClifford() && "LHS Operation is not Clifford");
+    assert(rhs.isNonClifford() && "RHS Operation is not non-Clifford");
+    assert(lhs.getPauliProduct().size() == lhs.getOutQubits().size() &&
+           "LHS Pauli product size mismatch before commutation.");
+    assert(rhs.getPauliProduct().size() == rhs.getInQubits().size() &&
+           "RHS Pauli product size mismatch before commutation.");
+
+    // Update Pauli words of RHS
+    if (result != nullptr) {
+        updatePauliWord(rhs, result->get_pauli_word(), rewriter);
+        updatePauliWordSign(rhs, result->isNegative(), rewriter);
+    }
+    else {
+        updatePauliWord(rhs, rhsPauli.get_pauli_word(), rewriter);
+        updatePauliWordSign(rhs, rhsPauli.isNegative(), rewriter);
+    }
```

Comment:
One last thing, wasn't the else branch superfluous here?
```suggestion
    // Update Pauli words of RHS
    if (result != nullptr) {
        updatePauliWord(rhs, result->get_pauli_word(), rewriter);
        updatePauliWordSign(rhs, result->isNegative(), rewriter);
    }
```

Reply:
Hi @dime10. I've just checked that function. This `else` branch is needed because the Pauli word may be reordered after normalizing.